src/Domain/SegmentId.hpp

```diff
@@ -0,0 +1,114 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines class SegmentId.
+
+#pragma once
+
+#include <cstdint>
+#include <iosfwd>
+#include <pup.h>
+
+#include "Domain/Side.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/ConstantExpressions.hpp"
+
+/// \ingroup ComputationalDomain
+/// A SegmentId labels a segment of the interval [-1,1].
+/// A SegmentId is used to (partially) identify an Element.
+/// The segment is the interval [-1 + 2*(N/D), -1 + 2*(N+1)/D]
+/// where N = index_ and D = 2 ^ refinement_level_.
+class SegmentId {
+ public:
+  /// Default constructor needed for Charm++ serialization.
+  SegmentId();
+
+  /// Construct from the refinement level and an index.
+  SegmentId(size_t refinement_level, size_t index);
+
+  /// The refinement level.
+  size_t refinement_level() const { return refinement_level_; }
+
+  /// The index of the segement on the refinement level.
+  size_t index() const { return index_; }
+
+  /// The id of the parent of this segment.
+  SegmentId id_of_parent() const;
+
+  /// The id of the child of this segment on the given Side.
+  SegmentId id_of_child(Side side) const;
+
+  /// The id this segment would have if the coordinate axis were flipped.
+  SegmentId id_if_flipped() const;
+
+  /// The logical coordinate of the endpoint of the segment on the given Side.
+  double endpoint(Side side) const;
+
+  /// The logical coordinate of the midpoint of the segment
+  double midpoint() const {
+    return -1.0 + (1.0 + 2.0 * index_) / two_to_the(refinement_level_);
+  }
+
+  /// Serialization for Charm++
+  void pup(PUP::er& p);  // NOLINT
+
+ private:
+  size_t refinement_level_;
+  size_t index_;
+};
+
+/// Output operator for SegmentId.
+std::ostream& operator<<(std::ostream& os, const SegmentId& id);
+
+/// Equivalence operator for SegmentId.
+bool operator==(const SegmentId& lhs, const SegmentId& rhs);
+
+/// Inequivalence operator for SegmentId.
+bool operator!=(const SegmentId& lhs, const SegmentId& rhs);
+
+//##############################################################################
+// INLINE DEFINITIONS
+//##############################################################################
+
+inline SegmentId SegmentId::id_of_parent() const {
+  ASSERT(0 != refinement_level_, "Called on root refinement level!");
+  return SegmentId(refinement_level_ - 1, index_ >> 1);
+}
+
+inline SegmentId SegmentId::id_of_child(Side side) const {
+  // We cannot ASSERT on the maximum level because it's only known at runtime
+  // and only known elsewhere in the code, not by SegmentId. I.e. SegmentId is
+  // too low-level to know about this.
+  if (Side::Lower == side) {
+    return SegmentId(refinement_level_ + 1, index_ << 1);
```

Comment:
Maybe I'm just dumb but I find it hard to parse bit manipulations quickly, I think anywhere we do something like this we should comment why.